Block Support: Add height block support feature

aaronrobertshaw opened the pull request:
Related: https://github.com/WordPress/gutenberg/issues/28356
Depends on: https://github.com/WordPress/gutenberg/pull/32392

## Description
Adds height block support and includes in new Dimensions panel.

## How has this been tested?
Theme.json test:
`wp-env run phpunit 'phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist --verbose --filter stylesheet /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php'`

Style.js test:
`npm run test-unit:watch packages/block-editor/src/hooks/test/style.js`

#### Manual Test Instructions

1. Checkout this PR and build
2. Enable custom dimensions support including height, padding and margin - [theme.json gist](https://gist.github.com/aaronrobertshaw/5556efc230d25f337d9ffe5ff9580e2d)
3. Enable dimensions block support for Group block - [block.json gist](https://gist.github.com/aaronrobertshaw/3532aa40d7b154ffdef720e7d249705a)
4. Create a post, add a group block and select it
5. Ensure dimensions panel appears in the sidebar
6. Open the dimension panel's more menu and toggle on the height control
7. Adjust the height can confirm the block's height gets adjusted. Inspect element in dev tools to confirm CSS rule
8. Save post and view on the frontend. The height style should be applied correctly.
9. Repeat the process for a dynamic block, ensuring correct styles on the frontend.
10. Switch to the site editor, open the global styles sidebar and select the `By Block Type` tab.
11. Confirm the global styles dimensions panel includes height for the group block and that it works.

## Screenshots <!-- if applicable -->
<img src="https://user-images.githubusercontent.com/60436221/121115463-74f7b480-c858-11eb-94a8-2d7db9ed0da8.gif" width="500"/>

## Types of changes
New feature.

## Next Steps

- [x] Possibly add min/max height support alongside this? 
PR for min-height: https://github.com/WordPress/gutenberg/pull/33970

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jasmussen commented:
Very nice. Love the progressive panel, and the height control doesn't suffer from the same challenge that [the width control](https://github.com/WordPress/gutenberg/pull/32502#issuecomment-856556381) does.

ramonjd commented:
> Possibly add min/max height support alongside this?

This would be a great addition @aaronrobertshaw. I thinking specifically about the cover block mentioned in this issue: https://github.com/WordPress/gutenberg/issues/23770#issuecomment-895703471

I could try to tackle adding it if you haven't already. It would mean adding two new hooks similar to what you've done in the PR I assume for now, in the absence of any designs consolidating width/min-width/max-width and so on...?

aaronrobertshaw commented:
> I could try to tackle adding it if you haven't already. It would mean adding two new hooks similar to what you've done in the PR I assume for now, in the absence of any designs consolidating width/min-width/max-width and so on...?

I haven't started on min/max heights or widths yet. I was holding off until I got some more traction on the primary height/width block supports.

It would be great if you'd like to pick those up. The sooner they are in place the smoother we can transition the Cover block's ad hoc min-height to use all block support features and unify its spacing and dimensions panels.

oandregal commented on lib/block-supports/dimensions.php:
```diff
@@ -21,10 +21,10 @@ function gutenberg_register_dimensions_support( $block_type ) {
 
-	$has_spacing_support = gutenberg_block_has_support( $block_type, array( 'spacing' ), false );
-	// Future block supports such as height & width will be added here.
+	$has_dimensions_support = gutenberg_block_has_support( $block_type, array( '__experimentalDimensions' ), false );
```
Sorry for being late for the conversation. I've noticed that in https://github.com/WordPress/gutenberg/pull/32392 we renamed a lot of things from spacing to dimensions. Couldn't look at it deeply yet, but wanted to flag that changing keys in `block.json` or `theme.json` is problematic if they already landed in WordPress core.

For example, here we're changing `spacing` to `__experimentalDimensions` in `block.json`. But `spacing` has already landed in WordPress core, so we need to keep `spacing` working as before. If we absolutely need to rename it, we can add some code that converts `spacing` into `dimensions`. If it's just a name preference, I'd advise to keep using what we have (spacing) and don't make our work harder for this.

At the UI level, it can be labeled differently if we need to and we also can change it as many times as we need.

FYI @jorgefilipecosta @youknowriad

oandregal replied:
I've gone through a first pass at #32392 I may have found an issue https://github.com/WordPress/gutenberg/pull/32392/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.php#r686752790

If we keep the `spacing` key, I'd suggest we rename the things back to spacing internally, although at the UI level we name them "dimensions".

aaronrobertshaw replied:
> Couldn't look at it deeply yet, but wanted to flag that changing keys in block.json or theme.json is problematic if they already landed in WordPress core.

Thanks @nosolosw, I'd thought of that and thought I'd maintained that backwards compatibility. Happy to change whatever I've messed up. I'm missing where that is though.

> For example, here we're changing spacing to __experimentalDimensions in block.json. But spacing has already landed in WordPress core, so we need to keep spacing working as before.

I've actually not changed `spacing` to `__experimentalDimensions` in the block.json or theme.json. In the line below this comment you can still see the check for the `spacing` portion of the overall dimensions themed block support.

My plan was to introduce the new specific dimensions related supports under "dimensions" itself. It didn't make sense to me to have `height`, `width` etc under `spacing`.

Some points about where the separation of dimensions and spacing keys/support occurs are:
- `lib/class-wp-theme-json-gutenberg.php` adds `dimensions` in addition to, rather than replacing `spacing`
- `lib/theme.json` adds the setting for `dimensions.customHeight` separate to `spacing`
- `packages/block-editor/src/hooks/dimensions.js` checks separate block.json flags for `spacing` vs `__experimentalDimensions` support
- `packages/block-editor/src/hooks/style.js` dimensions key was added alongside spacing for the inline style generation
- The padding/margin styles are still under `spacing` within the block attributes style object
- The written tests also indicate that the spacing keys were still maintained.

Your [comment](https://github.com/WordPress/gutenberg/pull/32392#discussion_r686752790) regarding the key used when registering the block support in `dimensions.php` was well made and I have a fix for that specifically in https://github.com/WordPress/gutenberg/pull/34030.

Is there anything else I have missed? Thanks for your help as always!

oandregal replied:
Thanks for the quick turnaround, Aaron. I have a [second thought](https://github.com/WordPress/gutenberg/pull/32392/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.php&hide-deleted-files=true#r687501190) about the other PR. I can review this one properly once we clarify whether what I've brought up is an issue or not.

aaronrobertshaw replied:
I've replied on the other PR [seeking a little clarification](https://github.com/WordPress/gutenberg/pull/32392#discussion_r687730001) on the desired changes. Hopefully, I'll have that fixed up tomorrow. Then will make the required changes here.

There was also some work based on this PR around [adding `min-height` support](https://github.com/WordPress/gutenberg/pull/33970) and using it to [unify two "dimensions" panels in the Cover block](https://github.com/WordPress/gutenberg/pull/34008). It looks like that may also need some further discussion although the hope was to have that sorted before the next release.

Really appreciate your insight on the dimensions/spacing backwards compatibility issue with core. Thanks!

aaronrobertshaw committed:
Add height dimensions block support

After the release of 5.9, the theme.json php class related changes will be added via a separate isolated commit.

aaronrobertshaw committed:
Initial updates to extend theme.json classes post 5.9

oandregal committed:
New settings (works in WordPress 5.9)

Rename wp_get_global_settings to gutenberg_get_global_settings.

In WordPress 5.9 the wp_* function is already defined,
so we can't override them. It's calling the existing
WP_Theme_JSON classes in WordPress core so it won't pick up
the plugin modification.

In the plugin, to make sure these changes work fine in 5.9
as well, we need to use the gutenberg_* function instead.

oandregal committed:
New styles (works in WordPress 5.9)

oandregal committed:
Fix for WordPress 5.9: gutenberg_get_global_stylesheet should call the Gutenberg classes

oandregal committed:
Fix lint issues

oandregal committed:
Move functions to the compat/6.0 folder

While this PR does not do any behavioral change,
we still need to call the plugin clasess when the
minimum plugin version is 5.9 (at which point,
the compat/5.9 is removed).

aaronrobertshaw committed:
Add `get_theme_item` to global styles controller

aaronrobertshaw commented:
I've cherry-picked some commits from https://github.com/WordPress/gutenberg/pull/38883 to allow the proper extension of theme.json and global styles. 

At this time, this PR is working however there has been a recent change of plans around how we might manage the general layout of blocks. This might actually include new layout blocks specifically. It is yet to be seen how appropriate control over height and width will be within this context. 

As a result, I will close this PR and the ones related to width support (https://github.com/WordPress/gutenberg/pull/32502, https://github.com/WordPress/gutenberg/pull/32878)
